src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/LocalDBAPI.cs

```diff
@@ -10,25 +10,37 @@ namespace Microsoft.Data
 {
     internal static partial class LocalDBAPI
     {
-        private const string const_localDbPrefix = @"(localdb)\";
+        private const string LocalDbPrefix = @"(localdb)\";
+        private const string LocalDbPrefix_NP = @"np:\\.\pipe\LOCALDB#";
 
 
         [UnmanagedFunctionPointer(CallingConvention.Cdecl, CharSet = CharSet.Unicode)]
         private delegate int LocalDBFormatMessageDelegate(int hrLocalDB, uint dwFlags, uint dwLanguageId, StringBuilder buffer, ref uint buflen);
 
         // check if name is in format (localdb)\<InstanceName - not empty> and return instance name if it is
+        // localDB can also have a format of np:\\.\pipe\LOCALDB#<some number>\tsql\query
         internal static string GetLocalDbInstanceNameFromServerName(string serverName)
         {
-            if (serverName == null)
-                return null;
-            serverName = serverName.TrimStart(); // it can start with spaces if specified in quotes
-            if (!serverName.StartsWith(const_localDbPrefix, StringComparison.OrdinalIgnoreCase))
-                return null;
-            string instanceName = serverName.Substring(const_localDbPrefix.Length).Trim();
-            if (instanceName.Length == 0)
-                return null;
-            else
-                return instanceName;
+            if (serverName is not null)
+            {
+                // it can start with spaces if specified in quotes
+                // Memory allocation is reduced by using ReadOnlySpan
+                ReadOnlySpan<char> input = serverName.AsSpan().Trim();
+                if (input.StartsWith(LocalDbPrefix.AsSpan(), StringComparison.OrdinalIgnoreCase))
+                {
+                    input = input.Slice(LocalDbPrefix.Length);
+                    if (!input.IsEmpty)
+                    {
+                        return input.ToString();
+                    }
+                }
+                else if (input.StartsWith(LocalDbPrefix_NP.AsSpan(), StringComparison.OrdinalIgnoreCase))
+                {
+                    return input.ToString();
```

Comment:
I think the input should also be checked if it's empty here. I'm getting different errors when using the prefixes as the server:

`(localdb)\////`;
(provider: SNI_PN11, error: 51 - An instance name was not specified while connecting to a Local Database Runtime. Specify an instance name in the format (localdb)\instance_name.)

`np:\\.\pipe\LOCALDB#/`:
(provider: Named Pipes Provider, error: 40 - Could not open a connection to SQL Server)

Reply:
this is user error. The error is pretty much explanatory and that is the behavior from before for (localdb).
1- (localdb)\.
2-(localdb)\Instance name
3- (localdb)\.\shared instance name

Reply:
for named pipe:
the only part changing in `np:\\.\pipe\LOCALDB#SH452F48\tsql\query` is the number inside it

Reply:
we could use other methods such as RegEx, but that would be an overkill. The fact that connection string starts with `np:\\.\pipe\LOCALDB` is enough to conclude that user is trying LocalDB

Reply:
In the case of `(localdb)\` , we tell the user that the server name is not in the correct format, i.e. the instance name is missing. In the case of `np:\\.\pipe\LOCALDB#/`, shouldn't we also tell them that the instance name is missing?

Reply:
> In the case of `(localdb)\` , we tell the user that the server name is not in the correct format, i.e. the instance name is missing. In the case of `np:\\.\pipe\LOCALDB#/`, shouldn't we also tell them that the instance name is missing?

I dont think so. Because that pattern is not created by user. It is in a folder in LocalDB inside Program File folder. So, you cant just create whatever you want. One can read the name from the folder using a process or copy paste it, whereas (localdb) could have some user mistakes. Plus are you testing in managed SNI?

Reply:
Yes, it is the same in managed SNI.

The whole point of error handling is that we can't assume that the user knows what they're doing and got their named pipe server name from the correct source. If we assumed that the user does the right thing every time, there would be no need for error handling.

The error message when trying to connect to `np:\\.\pipe\LOCALDB#/` states that a connection could not be opened because the server was not found or was inaccessible. We should be more specific and say that the connection failed because the provided server name was not even in the correct format.

Reply:
I see merit in aligning the errors produced between tcp and np scenarios. However, I wouldn't block this PR since we are not changing the behavior here. The behavior can be addressed in the future so that we don't delay fixing this issue for the upcoming hotfix.